File: homework3/Csharp/Program5.cs
using System;
using System.Collections.Generic;

namespace Homework3.Csharp
{
    // Task 5: Applied Enumeration - Order processing pipeline
    internal static class Program5
    {
        private enum OrderStatus
        {
            Pending,
            Confirmed,
            Processing,
            Shipped,
            Delivered,
            Cancelled,
            Refunded
        }

        private sealed class Order
        {
            public string OrderId { get; }
            public IReadOnlyList<string> Items { get; }
            public decimal Total { get; }
            public OrderStatus Status { get; private set; }
            public string? TrackingNumber { get; private set; }

            public Order(string orderId, IReadOnlyList<string> items, decimal total)
            {
                OrderId = orderId;
                Items = items;
                Total = total;
                Status = OrderStatus.Pending;
            }

            public bool CanTransitionTo(OrderStatus newStatus)
            {
                // Enum-driven state machine for orders.
                var validTransitions = new Dictionary<OrderStatus, OrderStatus[]>
                {
                    { OrderStatus.Pending,   new[] { OrderStatus.Confirmed, OrderStatus.Cancelled } },
                    { OrderStatus.Confirmed, new[] { OrderStatus.Processing, OrderStatus.Cancelled } },
                    { OrderStatus.Processing,new[] { OrderStatus.Shipped, OrderStatus.Cancelled } },
                    { OrderStatus.Shipped,   new[] { OrderStatus.Delivered } },
                    { OrderStatus.Delivered, new[] { OrderStatus.Refunded } },
                    { OrderStatus.Cancelled, Array.Empty<OrderStatus>() },
                    { OrderStatus.Refunded,  Array.Empty<OrderStatus>() }
                };

                OrderStatus[] allowedNextStates = validTransitions.TryGetValue(Status, out OrderStatus[]? allowed)
                    ? allowed
                    : Array.Empty<OrderStatus>();

                foreach (OrderStatus candidate in allowedNextStates)
                {
                    if (candidate == newStatus)
                    {
                        return true;
                    }
                }

                return false;
            }

            public bool TransitionTo(OrderStatus newStatus)
            {
                if (!CanTransitionTo(newStatus))
                {
                    return false;
                }

                // Enum-specific side effects
                switch (newStatus)
                {
                    case OrderStatus.Confirmed:
                        Console.WriteLine($"\tSending confirmation email for order {OrderId}");
                        break;
                    case OrderStatus.Processing:
                        Console.WriteLine("\tProcessing payment and preparing items");
                        break;
                    case OrderStatus.Shipped:
                        TrackingNumber = $"TRACK-{OrderId}-{GetHashCode()}";
                        Console.WriteLine($"\tOrder shipped! Tracking: {TrackingNumber}");
                        break;
                    case OrderStatus.Delivered:
                        Console.WriteLine("\tOrder delivery confirmed, thank you!");
                        break;
                    case OrderStatus.Refunded:
                        Console.WriteLine($"\tRefund processed: ${Total:F2}");
                        break;
                    case OrderStatus.Cancelled:
                        Console.WriteLine("\tOrder cancelled");
                        break;
                }

                Status = newStatus;
                return true;
            }

            public string GetCustomerMessage()
            {
                return Status switch
                {
                    OrderStatus.Pending => "Your order is pending. Please confirm.",
                    OrderStatus.Confirmed => "Order confirmed! We'll process it soon.",
                    OrderStatus.Processing => "Your order is being prepared...",
                    OrderStatus.Shipped => $"Your order shipped! Tracking: {TrackingNumber}",
                    OrderStatus.Delivered => "Your order was delivered. Leave a review!",
                    OrderStatus.Cancelled => "Your order has been cancelled.",
                    OrderStatus.Refunded => $"Your refund of ${Total:F2} has been processed.",
                    _ => "Unknown status"
                };
            }

            public void DisplayStatus()
            {
                Console.WriteLine($"Order {OrderId}: {Status.ToString().ToUpperInvariant()}");
                Console.WriteLine($"\tItems: {string.Join(", ", Items)}");
                Console.WriteLine($"\tTotal: ${Total:F2}");
                Console.WriteLine($"\tMessage: {GetCustomerMessage()}");
            }
        }

        private static void Main()
        {
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("Order Processing Pipeline");
            Console.WriteLine(new string('=', 70));
            Console.WriteLine();
            Console.WriteLine("Demonstrating how enum values drive the logic:");
            Console.WriteLine();

            var order = new Order(
                orderId: "ORD-12345",
                items: new[] { "Laptop", "Mouse", "Keyboard" },
                total: 1299.99m);

            Console.WriteLine($"Created order {order.OrderId}");
            order.DisplayStatus();

            var transitionsToTry = new[]
            {
                OrderStatus.Confirmed,
                OrderStatus.Processing,
                OrderStatus.Shipped,
                OrderStatus.Delivered,
                OrderStatus.Refunded
            };

            Console.WriteLine();
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("Processing Valid State Transitions:");
            Console.WriteLine(new string('=', 70));

            foreach (OrderStatus targetStatus in transitionsToTry)
            {
                Console.WriteLine();
                Console.WriteLine($"Attempt transition: {order.Status} -> {targetStatus}");
                bool ok = order.TransitionTo(targetStatus);
                Console.WriteLine(ok ? " Transition successful" : " Invalid transition");
                order.DisplayStatus();
            }

            Console.WriteLine();
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("Attempting Invalid Transition:");
            Console.WriteLine(new string('=', 70));

            Console.WriteLine();
            Console.WriteLine($"Order is in state: {order.Status}");
            Console.WriteLine($"Attempt transition: {order.Status} -> {OrderStatus.Processing}");
            bool invalid = order.TransitionTo(OrderStatus.Processing);
            Console.WriteLine(invalid ? " Transition successful" : " Invalid transition (already delivered/refunded)");

            Console.WriteLine();
            Console.WriteLine(new string('=', 70));
            Console.WriteLine("Alternative Path: Order Cancellation");
            Console.WriteLine(new string('=', 70));

            var order2 = new Order(
                orderId: "ORD-54321",
                items: new[] { "Book" },
                total: 29.99m);

            Console.WriteLine();
            Console.WriteLine($"Created order {order2.OrderId}");
            order2.DisplayStatus();

            var cancelPath = new[]
            {
                OrderStatus.Confirmed,
                OrderStatus.Cancelled
            };

            foreach (OrderStatus targetStatus in cancelPath)
            {
                Console.WriteLine();
                Console.WriteLine($"Attempt transition: {order2.Status} -> {targetStatus}");
                bool ok = order2.TransitionTo(targetStatus);
                Console.WriteLine(ok ? " Transition successful" : " Invalid transition");
            }
        }
    }
}

File: homework3/Python/Program5.py
from enum import Enum
from typing import Optional

# THis is an enum that tracks different Order Statuses of an e-commerce platform. 
# These are predefined by the client and/or business rules
class OrderStatus(Enum):

    PENDING = "pending"
    CONFIRMED = "confirmed"
    PROCESSING = "processing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    REFUNDED = "refunded"

# This class defines what an Order i and how it behaves. First we build a constructor of what can an order have
# then, we create a chart that defines what can a status transition so this means that each orderstatus is dependent on
# other orderstatuses. Then we created a function that transition the status of the order to the new status depending on the chart and user. 
# There is also another function that gives a predetermined message to the user depending on the order status. 
# Finally, we have a function that displays the order status and other information about the order.
class Order:
 
    def __init__(self, orderID: str, items: list, total: float):
        self.orderID = orderID
        self.items = items
        self.total = total
        self.status: OrderStatus = OrderStatus.PENDING
        self.orderTrackingNumber: Optional[str] = None
    
    def orderTransitionChart(self,newOrderStatus: OrderStatus) -> bool:
       
        validTransitions = {
            OrderStatus.PENDING: [OrderStatus.CONFIRMED, OrderStatus.CANCELLED],
            OrderStatus.CONFIRMED: [OrderStatus.PROCESSING, OrderStatus.CANCELLED],
            OrderStatus.PROCESSING: [OrderStatus.SHIPPED, OrderStatus.CANCELLED],
            OrderStatus.SHIPPED: [OrderStatus.DELIVERED],
            OrderStatus.DELIVERED: [OrderStatus.REFUNDED],
            OrderStatus.CANCELLED: [],  
            OrderStatus.REFUNDED: [],   
        }
        
        allowedNextStates = validTransitions.get(self.status, [])
        return newOrderStatus in allowedNextStates
    
    def transitionFunction(self, newOrderStatus: OrderStatus) -> bool:
      
        if not self.orderTransitionChart(newOrderStatus):
            return False
        if newOrderStatus == OrderStatus.CONFIRMED:
            print(f"  → Sending confirmation email for order {self.orderID}")
        elif newOrderStatus == OrderStatus.PROCESSING:
            print(f"  → Processing payment and preparing items")
        elif newOrderStatus == OrderStatus.SHIPPED:
            self.orderTrackingNumber = f"TRACK-{self.orderID}-{id(self)}"
            print(f"  → Order shipped! Tracking: {self.orderTrackingNumber}")
        elif newOrderStatus == OrderStatus.DELIVERED:
            print(f"  → Order delivery confirmed, thank you!")
        elif newOrderStatus == OrderStatus.REFUNDED:
            print(f"  → Refund processed: ${self.total:.2f}")
        elif newOrderStatus == OrderStatus.CANCELLED:
            print(f"  → Order cancelled")
        
        self.status = newOrderStatus
        return True
    
    def getCustomerMessage(self) -> str:
        
        messages = {
            OrderStatus.PENDING: "Your order is pending. Please confirm.",
            OrderStatus.CONFIRMED: "Order confirmed! We'll process it soon.",
            OrderStatus.PROCESSING: "Your order is being prepared...",
            OrderStatus.SHIPPED: f"Your order shipped! Tracking: {self.orderTrackingNumber}",
            OrderStatus.DELIVERED: "Your order was delivered. Leave a review!",
            OrderStatus.CANCELLED: "Your order has been cancelled.",
            OrderStatus.REFUNDED: f"Your refund of ${self.total:.2f} has been processed.",
        }
        return messages.get(self.status, "Unknown status")
    
    def display_status(self) -> None:
       
        print(f"Order {self.orderID}: {self.status.value.upper()}")
        print(f"  Items: {', '.join(self.items)}")
        print(f"  Total: ${self.total:.2f}")
        print(f"  Message: {self.getCustomerMessage()}")



# The main function creates a predetermined order with information, these are just examples. 
# In main the order goes through different steps of the processing pipeline where it emulates a real processing of an order.
# an order is created -> changes from pending to confirmed -> processing -> shipped -> delivered -> refunded. 
# Then we try to make an invalid transition and we see that it is not possible. 
# Finally, we create another order and we try to cancel it.
# So, the main idea is that without enum, the logic has to be different since we would have to write more code to check for the different states
# Also, we would have to write many ways to safely transition between states, but with enums we can just define the valid
# states that a state can transition to
# Removing enum will overall lead to a huge refactoring that it can be avoided with the use of enums. 
def main():
    print("REAL-WORLD ENUM APPLICATION: Order Processing Pipeline")
    print("\nDemonstrating how enum values drive the logic:\n")
    
    order = Order(
        orderID="ORD-12345",
        items=["Laptop", "Mouse", "Keyboard"],
        total=1299.99
    )
    
    print(f"Created order {order.orderID}")
    order.display_status()

    transitionToTry = [
        OrderStatus.CONFIRMED,
        OrderStatus.PROCESSING,
        OrderStatus.SHIPPED,
        OrderStatus.DELIVERED,
        OrderStatus.REFUNDED,
    ]
    
    print("\n" )
    print("Processing Valid State Transitions:")
    
    for target_status in transitionToTry:
        print(f"\nAttempt transition: {order.status.value} → {target_status.value}")
        if order.transitionFunction(target_status):
            print(f"✓ Transition successful")
        else:
            print(f"✗ Invalid transition")
        order.display_status()
    
  
    print("\n" )
    print("Attempting Invalid Transition:")
    
    print(f"\nOrder is in state: {order.status.value}")
    print(f"Attempt transition: {order.status.value} → {OrderStatus.PROCESSING.value}")
    if order.transitionFunction(OrderStatus.PROCESSING):
        print(f"Transition successful")
    else:
        print(f"Invalid transition (already delivered/refunded)")
    
    print("\n" )
    print("Alternative Path: Order Cancellation")
    
    order2 = Order(
        orderID="ORD-54321",
        items=["Book"],
        total=29.99
    )
    
    print(f"\nCreated order {order2.orderID}")
    order2.display_status()
    
    cancel_path = [
        OrderStatus.CONFIRMED,
        OrderStatus.CANCELLED,
    ]
    
    for target_status in cancel_path:
        print(f"\nAttempt transition: {order2.status.value} → {target_status.value}")
        if order2.transitionFunction(target_status):
            print(f"Transition successful")
        else:
            print(f"Invalid transition")
    
    print("\n" )


main()
